src/main/scala/is/hail/table/Table.scala

```diff
@@ -626,70 +626,104 @@ class Table(val hc: HailContext, val ir: TableIR) {
-      .zipWithIndex
-      .map { case (fd, i) => (finalColumnNames(i), fd.typ) }: _*)
-    val localNKeys = nKeys
-    val size1 = valueSignature.size
-    val size2 = other.valueSignature.size
-    val totalSize = newSignature.size
+    val newSignature = TStruct(joinedFields.zipWithIndex.map {
+      case (fd, i) => (finalColumnNames(i), fd.typ)
+    }: _*)
+
+    val leftSignature = signature
+    val rightSignature = other.signature
+    val leftKeyFieldIdx = keyFieldIdx
+    val rightKeyFieldIdx = other.keyFieldIdx
+    val leftValueFieldIdx = valueFieldIdx
+    val rightValueFieldIdx = other.valueFieldIdx
+    val rvMerger: Iterator[JoinedRegionValue] => Iterator[RegionValue] = { it =>
+      val rvb = new RegionValueBuilder()
+      val rv = RegionValue()
+      it.map { joined =>
+        val lrv = joined._1
+        val rrv = joined._2
+
+        if (lrv != null)
+          rvb.set(lrv.region)
+        else {
+          assert(rrv != null)
+          rvb.set(rrv.region)
+        }
 
-    assert(totalSize == localNKeys + size1 + size2)
+        rvb.start(newSignature)
+        rvb.startStruct()
 
-    val merger = (k: Row, r1: Row, r2: Row) => {
-      val result = Array.fill[Any](totalSize)(null)
+        if (lrv != null)
+          rvb.addFields(leftSignature, lrv, leftKeyFieldIdx)
+        else {
+          assert(rrv != null)
+          rvb.addFields(rightSignature, rrv, rightKeyFieldIdx)
+        }
 
-      var i = 0
-      while (i < localNKeys) {
-        result(i) = k.get(i)
-        i += 1
-      }
+        if (lrv != null)
+          rvb.addFields(leftSignature, lrv, leftValueFieldIdx)
+        else
+          rvb.skipFields(leftValueFieldIdx.length)
 
-      if (r1 != null) {
-        i = 0
-        while (i < size1) {
-          result(localNKeys + i) = r1.get(i)
-          i += 1
-        }
-      }
+        if (rrv != null)
+          rvb.addFields(rightSignature, rrv, rightValueFieldIdx)
+        else
+          rvb.skipFields(rightValueFieldIdx.length)
 
-      if (r2 != null) {
-        i = 0
-        while (i < size2) {
-          result(localNKeys + size1 + i) = r2.get(i)
-          i += 1
-        }
+        rvb.endStruct()
+        rv.set(rvb.region, rvb.end())
+        rv
       }
-      Row.fromSeq(result)
     }
 
-    val rddLeft = keyedRDD()
-    val rddRight = other.keyedRDD()
-
-    val joinedRDD = joinType match {
-      case "left" => rddLeft.leftOuterJoin(rddRight).map { case (k, (l, r)) => merger(k, l, r.orNull) }
-      case "right" => rddLeft.rightOuterJoin(rddRight).map { case (k, (l, r)) => merger(k, l.orNull, r) }
-      case "inner" => rddLeft.join(rddRight).map { case (k, (l, r)) => merger(k, l, r) }
-      case "outer" => rddLeft.fullOuterJoin(rddRight).map { case (k, (l, r)) => merger(k, l.orNull, r.orNull) }
-      case _ => fatal("Invalid join type specified. Choose one of `left', `right', `inner', `outer'")
+    val left = this.rvd match {
+      case ordered: OrderedRVD => ordered
```

Comment:
what about the case that it's ordered but keyed by the wrong thing? I'm having a bit of trouble reasoning about that case right now. Could it happen?

Reply:
ah, keyBy does a horrible thing right now. so maybe it can't?

Reply:
let's add an assertion, at least.

Reply:
I think we need to decide what keys mean in different places. One thing I think would be reasonable is
* Keys on `Table`, both in Python and in Scala, are purely metadata used to determine how to join.
* Keys on `OrderedRVD` are a way of recording the ordering invariant satisfied by the underlying `RDD`.

If this is the case, then I need to rewrite this. When I wrote it, I think I was assuming the `Table` keys and the `OrderedRVD` keys had to be the same. Now I understand `Table`s are never ordered currently. We should probably make a reasonable choice of what keys mean and document it.